.github/workflows/perf-testing-setup.yml

```diff
@@ -47,13 +38,63 @@ jobs:
         python setup.py extension
     - name: Build and Run the Docker image
       run: |
+        echo "Building Docker image with Python version ${{ env.PYTHON_VERSION }}"
         docker build --build-arg PYTHON_VERSION=${{ env.PYTHON_VERSION }} --file .ci/perf_tests/dockerfiles/perf_tests.Dockerfile --tag perfimage:latest .
-        docker run -d --shm-size="2g" --env FUNCTIONS_WORKER_RUNTIME_VERSION=${{ env.PYTHON_VERSION }} -p ${PORT}:80 -v $GITHUB_WORKSPACE/azure_functions_worker:/azure-functions-host/workers/python/${{ env.PYTHON_VERSION }}/LINUX/X64/azure_functions_worker perfimage:latest
+    
+        echo "Running Docker container..."
+        container_id=$(docker run -d --privileged --shm-size="2g" --env FUNCTIONS_WORKER_RUNTIME_VERSION=${{ env.PYTHON_VERSION }} -p ${PORT}:80 -v $GITHUB_WORKSPACE/azure_functions_worker:/azure-functions-host/workers/python/${{ env.PYTHON_VERSION }}/LINUX/X64/azure_functions_worker perfimage:latest)
         sleep 10  # host needs some time to start.
+        echo "Container ID is $container_id"
+        
+        echo "Fetching Docker container logs..."
+        docker logs $container_id
+        worker_pid=$(docker exec $container_id sh -c "ps aux | grep '[p]ython'" | awk '{print $2}')
+        echo "Python worker process id is $worker_pid"
+        
+        echo "container_id=$container_id" >> $GITHUB_ENV
+        echo "worker_pid=$worker_pid" >> $GITHUB_ENV
+        
     - name: Validate if the functions are now running
       run: |
         curl --get http://localhost:${PORT}/api/${{ matrix.test_to_run }}
+        
+    - name: Start py-spy in the background
+      run: |
+        docker exec $container_id sh -c "pip install py-spy"
+        docker exec $container_id sh -c "mkdir /home/profiling_reports"
+        report_name="azpyfuncpprofiling_${{ github.run_id }}.svg"
+        docker exec -d $container_id sh -c "py-spy record -p $worker_pid -o /home/profiling_reports/$report_name -f flamegraph --idle --nonblocking --rate 1500 > /home/site/wwwroot/py-spy.log 2>&1 &"
+        sleep 2  # Give it a moment to start
+        py_spy_id=$(docker exec $container_id sh -c "ps aux | grep '[p]y-spy record'" | awk '{print $2}')
+        
+        echo "py_spy_id=$py_spy_id" >> $GITHUB_ENV
+        echo "report_name=$report_name" >> $GITHUB_ENV
+        
     - name: Run Throughput tests
       run: |
         chmod 755 .ci/perf_tests/run-perftests.sh
         .ci/perf_tests/run-perftests.sh localhost $PORT ${{ env.TESTS_DIR_PATH }} ${{ matrix.test_to_run }}
+    
+    - name: Stop profiling and generate report
```

Comment:
Do upload it as an artifact as well of the run

Reply:
updated